File: samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/configuration/SampleTopology.java
package systems.zlink.samples.zoneworld.server.configuration;

import org.springframework.boot.context.properties.ConfigurationProperties;

import java.util.Objects;

@ConfigurationProperties("sample")
public record SampleTopology(
        String role,
        String nodeId,
        String meshEndpoint,
        String streamEndpoint,
        String redisEndpoint,
        String redisKeyPrefix,
        Boolean subscriberOnly,
        Boolean disableBots,
        Boolean allowEmptyZoneSet,
        String faultTickZone,
        String meshAdvertiseHost) {

    public boolean is(String expected) {
        return expected.equalsIgnoreCase(role);
    }

    public boolean isSubscriberOnly() {
        return Boolean.TRUE.equals(subscriberOnly);
    }

    public boolean botsDisabled() {
        return Boolean.TRUE.equals(disableBots);
    }

    public boolean allowsEmptyZoneSet() {
        return Boolean.TRUE.equals(allowEmptyZoneSet);
    }

    public static String configPath(String[] args) {
        if (args.length != 2 || !"--config".equals(args[0]) || args[1].isBlank()) {
            throw new IllegalArgumentException("Usage: ZoneWorldServer --config <path>");
        }
        return args[1];
    }

    public void validateServer() {
        require(role, "sample.role");
        require(meshEndpoint, "sample.mesh-endpoint");
        require(redisEndpoint, "sample.redis-endpoint");
        require(redisKeyPrefix, "sample.redis-key-prefix");
        if (is("zone")) require(nodeId, "sample.node-id");
        if (is("gateway") || is("ops")) require(streamEndpoint, "sample.stream-endpoint");
        if (!is("zone") && !is("gateway") && !is("ops")) {
            throw new IllegalArgumentException("sample.role must be gateway, zone, or ops");
        }
    }

    private static void require(String value, String name) {
        if (Objects.requireNonNullElse(value, "").isBlank()) {
            throw new IllegalArgumentException(name + " is required");
        }
    }
}

File: samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/server/configuration/SampleTopology.kt
package systems.zlink.samples.kotlin.zoneworld.server.configuration

import org.springframework.boot.context.properties.ConfigurationProperties

@ConfigurationProperties("sample")
data class SampleTopology(
    val role: String? = null,
    val nodeId: String? = null,
    val meshEndpoint: String? = null,
    val streamEndpoint: String? = null,
    val redisEndpoint: String? = null,
    val redisKeyPrefix: String? = null,
    val subscriberOnly: Boolean? = null,
    val disableBots: Boolean? = null,
    val allowEmptyZoneSet: Boolean? = null,
    val faultTickZone: String? = null,
    val meshAdvertiseHost: String? = null,
) {
    fun isRole(expected: String) = role.equals(expected, ignoreCase = true)

    fun required(value: String?, name: String): String =
        value?.takeIf { it.isNotBlank() } ?: error("sample.$name is required")

    fun roleValue() = required(role, "role")

    fun nodeValue() = required(nodeId, "node-id")

    fun meshValue() = required(meshEndpoint, "mesh-endpoint")

    fun streamValue() = required(streamEndpoint, "stream-endpoint")

    fun redisValue() = required(redisEndpoint, "redis-endpoint")

    fun prefixValue() = required(redisKeyPrefix, "redis-key-prefix")

    fun isSubscriberOnly() = subscriberOnly == true

    fun botsDisabled() = disableBots == true

    fun allowsEmptyZoneSet() = allowEmptyZoneSet == true

    fun validate() {
        val value = roleValue()
        require(value == "zone" || value == "gateway" || value == "ops") {
            "sample.role must be gateway, zone, or ops"
        }
        if (value == "zone") nodeValue()
        if (value != "zone") streamValue()
        meshValue()
        redisValue()
        prefixValue()
    }

    companion object {
        fun configPath(args: Array<String>): String {
            require(args.size == 2 && args[0] == "--config" && args[1].isNotBlank()) {
                "Usage: ZoneWorldServer --config <path>"
            }
            return args[1]
        }
    }
}
